rfcs/0000-longjmp-pof-annotation.md

```diff
@@ -0,0 +1,85 @@
+- Feature Name: `annotation-for-safe-longjmp`
+- Start Date: 2019-06-11
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+- Project group: [FFI-unwind][project-group]
+
+ [project-group]: https://github.com/rust-lang/project-ffi-unwind
+
+ <!-- XXX for now, content is copied from prior drafts of the "C unwind" RFC. -->
+
+# Motivation
+[motivation]: #motivation
+
+Additionally, there are libraries such as `rlua` that rely on `longjmp` across
+Rust frames; on Windows, `longjmp` is implemented via [forced
+unwinding][forced-unwinding]. The current `rustc` implementation makes it safe
+to `longjmp` across Rust [POFs][POF-definition] (frames without `Drop` types),
+but this is not formally specified in an RFC or by the Reference.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This RFC specifies that, regardless of the platform or the ABI string (`"C"` or
+`"C unwind"`), any platform features that may rely on forced unwinding is
+defined behavior when all unwound frames are POFs
+
+As an example:
+
+```rust
+#[cancelable]
+fn foo<D: Drop>(c: bool, d: D) {
+  if c {
+    drop(d);
+  }
+  longjmp_if_true(c);
+}
+
+/// Calls `longjmp` if `c` is true; otherwise returns normally.
+extern "C" fn longjmp_if_true(c: bool);
+```
+
+If a `longjmp` occurs, it can safely traverse the `foo` frame, which will be a
+POF because `d` has already been dropped.
+
+Since `longjmp_if_true` function is using the `"C"` rather than the `"C
+unwind"` ABI, the optimizer may assume that it cannot unwind; on LLVM, this is
+represented by the `nounwind` attribute. On most platforms, `longjmp` is not a
+form of unwinding: the `foo` frame is simply discarded. On Windows, `longjmp`
+is implemented as a forced unwind, which is permitted to traverse `nounwind`
+frames. Since `foo` contains a `Drop` type the forced unwind will include a
+call to the frame's cleanup logic, but that logic will not produce any
+observable effect; in particular, `D::drop()` will not be called again. The
+observable behavior should therefore be the same on all platforms.
+
+Conversely, if, due to a bug, `longjmp` were called unconditionally, then this
+code would have undefined behavior on all platforms when `c` is false, because
+`foo` would not be a POF.

```

Comment:
I would expect that the compiler would warn in this scenario, also

Reply:
Maybe I'm misunderstanding, but how would the compiler provide a warning for this? The `extern` function is in C, so the compiler can't determine under what conditions it will `longjmp`. Or do you mean it would warn in any scenario where a frame is not guaranteed to be a POF at the time it calls a "cancelable" function?

Reply:
> do you mean it would warn in any scenario where a frame is not guaranteed to be a POF at the time it calls a "cancelable" function?

This is what I meant. I think it's going to be quite rare that you have a function that *sometimes cancels* but you happen to know that *on this particular call it will not* and therefore you want to take advantage of that by having a pending destructor. In such a case, adding an `#[allow]` annotation seems reasonable to me.

Reply:
Adding a comment to that effect